### DISCUSSION TITLE
DAG updates interference

### DISCUSSION THREAD

**Discussion by michmauzer004** [apache/airflow#36634] [Q&A]
How do updates to DAG affect currently running instances of these graphs, and what is the recommended practice for making changes to DAGs?

**Comment by y0zg** [comment]
When DAGs are updated, they do not interfere with currently running DAGs. But it’s preferable to make any changes when DAGs are not running

**Reply by potiuk** [reply]
This is wrong answer and looks like hallucinated AI generated response @y0zg - if you added this response based on some AI generating it then STOP doing it. I unmarked it as answer and please verify your answers when you post them and never post unverified AI answers.

Changing DAG when they are running **does** change the structure in somewhat unpredictable ways (it depends on races between dag file processors updating serialized DAG in the database, scheduler scanning it and scheduling next tasks. 

So you should avoid changing the structure too dramatically. DAG versioning Airflow Improvement Proposal https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning - is still in a draft stage and not actively worked on (for now) to address this. 

Best practice if you want to change DAG structure significantly for now is to copy your DAG file / code and create a new DAG with new dag_id to follow the new structure.

**Reply by y0zg** [reply]
Well, this is a note I have from airflow slack discussion, strange that you are referring to AI. I know regarding long pending DAG versioning feature which I don't think will be delivered in the future, this is airflow users pain. Moreover, I tested this, If we update docker image for already running DAG with several tasks, this affects only newly started DAG and not already running. Using Xcom could help to sync this misbehaviour

Example with k8s operator
```
import datetime
import pendulum
from airflow import models, DAG
from airflow.operators.python import PythonOperator
from airflow.operators.empty import EmptyOperator
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator
from kubernetes.client import models as k8s_models
from airflow.models import Variable, XCom

default_args = {
    'owner': 'airflow',
}

def fetch_variables():
    docker_image_task = Variable.get("docker_image_task")
    return {'docker_image_task': docker_image_task}

with DAG(
        dag_id="k8s-pod-xcom",
        default_args=default_args,
        schedule_interval=None,
        tags=["ex"],
        start_date=pendulum.datetime(2023, 7, 19)
    ) as dag:

    fetch_variables_task = PythonOperator(
        task_id='fetch_variables',
        python_callable=fetch_variables,
        dag=dag
    )

    start = EmptyOperator(task_id='run_this_first_repo_accounts', dag=dag)

    passing = KubernetesPodOperator(namespace='airflow',
                            image="{{ task_instance.xcom_pull(task_ids='fetch_variables')['docker_image_task'] }}",
                            cmds=["bash","-c"],
                            arguments=["echo 'task1 started' && sleep 30 && echo 'end of task1'"],
                            labels={"repo": "accounts"},
                            image_pull_policy="Always",
                            name="task-1",
                            task_id="passing-task",
                            get_logs=True,
                            )

    failing = KubernetesPodOperator(namespace='airflow',
                            image="{{ task_instance.xcom_pull(task_ids='fetch_variables')['docker_image_task'] }}",
                            cmds=["bash","-c"],
                            arguments=["echo 'task2 started' && sleep 30 && echo 'end of task2'"],
                            labels={"repo": "accounts"},
                            image_pull_policy="Always",
                            name="task-2",
                            task_id="failing-task",
                            get_logs=True,
                            )

    fetch_variables_task >> start >> passing >> failing
```

**Reply by potiuk** [reply]
Ok. Apologies then - I am overly sensitive on people posting here AI generated answers that makes no sense and are misleading people - it happened already several times by people who want to boost their "presence" in the project but do not check those - often hallicunated - answers and do not state the source of those. Stopping such people short is the only way we can prevent the forums from being flooded by wrong answers which make no sense.

And your answer suspiciously looked like one - sorry for that bad assesment.

Your example is fine but it only works on TASK level not DAG level.

What I was refererring to (and what I believe the question was about ) is what happens when DAG "structure" changes while DAGs are running - for example when you have a DAG with 100 tasks and you add 20 more with various dependencies between them (some of them already running. but some of them not yet even scheduled).

In this case what happens is a bit indeterministic.
